### ISSUE TITLE
[1.2.0] client.fetchNextPage() gives TypeError

### ISSUE THREAD

**Issue by rizedr** [Shopify/js-buy-sdk#493]
Getting error in `v1.2.0` with products fetched from a collection (error similar to the one described in #433):
1. `client.collection.fetchWithProducts(id) => products` works, got 20 products
2. `client.fetchNextPage(products) => products` works, got 40 products
3. `client.fetchNextPage(products) => products` works, got 60 products
4. `client.fetchNextPage(products) => error` doesn't work anymore.
```
Uncaught (in promise) TypeError: Cannot read property 'node' of undefined
    at index.js:2073
    at Array.reduce (<anonymous>)
    at index.js:2072
    at <anonymous>
```

I've made `create-react-app` to demonstrate the issue which can be found [here](https://github.com/rizedr/shopify-fetchNextPage) (can be cloned to reproduce the error locally). A deployed version is available to be seen here: http://shopify-next-page-bug.surge.sh/

**Comment by rizedr** [comment]
Can replicate this with [`graphql-js-client`](https://github.com/Shopify/graphql-js-client). This what I get for the [`graphql-js-client`'s fetchNextPage method `response`](https://github.com/Shopify/graphql-js-client/blob/master/src/client.js#L177):
```
{
  "errors": [
    {
      "message": "Parse error on \"$\" (VAR_SIGN) at [1, 21]",
      "locations": [
        {
          "line": 1,
          "column": 21
        }
      ]
    }
  ]
}
```

It has not `model` property, thus the `TypeError` happening. Any thoughts on why this is happening @minasmart? I can provide you with an `export.csv` on the products where this is happening.

**Comment by lisadreamer** [comment]
@rizedr have taken a look which parameter it writes error about? in my case graphQLParams.query looks like this 
"fragment ProductFragment on Product  { id,createdAt,updatedAt,descriptionHtml,description,handle,productType,title,vendor,tags,publishedAt,onlineStoreUrl,options { id,name,values },images (first: 250) { pageInfo { hasNextPage,hasPreviousPage },edges { cursor,node { id,src,altText } } },variants (first: 250) { pageInfo { hasNextPage,hasPreviousPage },edges { cursor,node { ...VariantFragment } } } },fragment VariantFragment on ProductVariant  { id,title,price,weight,available,sku,compareAtPrice,image { id,src,altText },selectedOptions { name,value } },query ($first:Int = $first,$query:String,$sortKey:ProductSortKeys,$reverse:Boolean)  { shop { products (first: $first query: $query sortKey: $sortKey reverse: $reverse after: "eyJsYXN0X2lkIjoyMzU1NDA3MDkzODUsImxhc3RfdmFsdWUiOiJTdW5zZXQgMDAxIn0=") { pageInfo { hasNextPage,hasPreviousPage },edges { cursor,node { ...ProductFragment } } } } }"

and my position error is at 577 which is $first inside of query ($first:Int = $first,$query:String,$sortKey:ProductSortKeys,$reverse:Boolean)

is it smth about rootQuery?

**Comment by rizedr** [comment]
@lisadreamer yes I have looked and the parameter that errors on is `$first` as well 👍

**Comment by lisadreamer** [comment]
so maybe defaultValue should be 20? for defaultValue of variableDefinition "first"...help us @minasmart 

btw @rizedr what r u using for first fetching? (fetchAll, fetchQuery)?

**Comment by rizedr** [comment]
@lisadreamer I **was** using `client.collection.fetchWithProducts`, but it shouldn't matter in terms of what method you are using to get the first list of products since all of them are returning a list of `GraphModel`s. 

Underneath, this module is using the `graphql-js-client`, thus, when you are doing `fetchNextPage`, it's using this [method](https://github.com/Shopify/graphql-js-client/blob/master/src/client.js#L177) which brings the error I pointed [here](https://github.com/Shopify/js-buy-sdk/issues/493#issuecomment-374042767).

The way I managed to avoid this was to dump `js-buy-sdk` all together and use `graphql-js-client` and implement my own queries. I followed the example in the [graphql react storefront api](https://github.com/Shopify/storefront-api-examples/tree/master/react-graphql-client). This is the query I've wrote for fetching the nextPage of products from a collection, but should be easy to update to fetch next pages from `shop`:
```javascript
const state = getState();
const products = getProducts(state);
const after = products[products.length - 1].cursor;
const id = getCollectionId(state, { handle });

const query = gql(graphClient)`
  query($id: ID!, $after: String!) {
    node(id: $id) {
      id
      ... on Collection {
        handle
        title
        products(first: 18, after: $after) {
          pageInfo {
            hasNextPage
            hasPreviousPage
          }
          edges {
            cursor
            node {
              id
              createdAt
              updatedAt
              descriptionHtml
              description
              handle
              productType
              title
              vendor
              tags
              publishedAt
              onlineStoreUrl
              options {
                name
                values
              }
              images(first: 250) {
                pageInfo {
                  hasNextPage
                  hasPreviousPage
                }
                edges {
                  cursor
                  node {
                    id
                    src
                    altText
                  }
                }
              }
              variants(first: 250) {
                pageInfo {
                  hasNextPage
                  hasPreviousPage
                }
                edges {
                  cursor
                  node {
                    id
                    title
                    price
                    weight
                    available
                    sku
                    compareAtPrice
                    image {
                      id
                      src
                      altText
                    }
                    selectedOptions {
                      name
                      value
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  }
`;

return graphClient.send(query, { id, after })
  .then((res) => {
    const cursors = res.data.node.products.edges;
    const nextProducts = res.model.node.products.map((item, index) => ({
      ...item,
      cursor: cursors[index].cursor,
    }));
    console.log('nextProducts', nextProducts);
  });
```

**Sadly**, I've run into another issue with I've opened (here)[https://github.com/Shopify/babel-plugin-graphql-js-client-transform/issues/13] - babal plugin that is necessary for you to use the template literals generated by `gql` package, don't know why is that needed since template literals are supported by default by ES6 but would love for someone from Shopify to reply!

**Comment by lisadreamer** [comment]
@rizedr i see..yeah i was feeling that i need to switch to graphql-js-client. but thought maybe im just using it wrong and i wanted to see if checkouts gonna work in every case I need...anyway i need to study first GraphQL better before )). but thanks for info. hope the other babel plugin issue gonna be solved.

**Comment by rizedr** [comment]
@lisadreamer go directly to the apollo-client or [else your app build will fail](https://github.com/Shopify/babel-plugin-graphql-js-client-transform/issues/13).

**Comment by luyx2412** [comment]
Same this issue in version 1.2.0
